More concise Follow-logs using aggregation

Anviking opened the pull request:
# Issue Number

ADP-630

# Overview

**1. Invert control to empower follow** 

The idea is: the more we can concentrate information in one place (follow), the better we can produce good logs.

- [x] Use `FollowLog tr` where `tr` can be `StakePoolLog` or `WalletWorkerLog`, instead of like now, having them both wrap `FollowLog`.
- [x] Move rollback-logic from call-sites to `follow`
   - [x] Ensure this commit _perfectly_ preserves previous behaviour -> should be pretty close

**2. Tweak logging**
- [x] Lower common messages to `Debug`
- [x] Add some kind of aggregation (I think doing manually is easier than learning how to use iohk-monitoring-framework for it, and maybe more powerful)
    - [x] Improve 
    - [x] Move `syncProgress` calculation to `follow` 
    - [x] Make sure we are strict where we need to, and don't create massive thunks


# Comments

- Based on PR #2581,  #2586 and #2615 branches - merge them first.

## Lost network connection + 5 min wait:

<img width="765" alt="Skärmavbild 2021-04-16 kl  16 58 45" src="https://user-images.githubusercontent.com/304423/115046205-7e5b4500-9ed7-11eb-945b-d098a855ba7b.png">

## On reconnection

<img width="774" alt="Skärmavbild 2021-04-20 kl  16 03 52" src="https://user-images.githubusercontent.com/304423/115409652-3b101780-a1f2-11eb-947d-eb3509b38919.png">


<!-- Additional comments or screenshots to attach if any -->

rvl commented on lib/core/src/Cardano/Wallet/Network.hs:
```diff
@@ -251,25 +263,80 @@ data FollowExit
+            case act of
+                FollowFailure ->
+                    -- TODO: Log here, e.g. MsgCrashMonitoring
+                    case recovery of
```
I would remove FollowFailure and let exceptions bubble up to the caller of `follow`.
The caller can decide whether to retry, possibly using the `recovering` function (which has good logging abilities).

Anviking replied:
As said elsewhere, my goal was to move as much responsibility from the callers to `follow`.

Maybe `recovering` (inside follow) is a good idea, but we didn't use it before at least, so we could keep it separate from this PR.

Anviking replied:
Btw, the pool worker simply stops if you kill and restart the node... We should probably either recover from exceptions for both pool and wallet workers, or re-add an `ExceptT` (#2581)

rvl commented on lib/core/src/Cardano/Wallet/Network.hs:
```diff
@@ -378,19 +459,90 @@ instance ToText FollowLog where
+
+--
+-- Log aggregation
+-- Perhaps this could be done using iohk-monitoring as well.
```
It can all be done with contra-tracer and tracer-transformers.

rvl commented on lib/core/src/Cardano/Wallet/Network.hs:
```diff
@@ -378,19 +459,90 @@ instance ToText FollowLog where
+-- - If the node goes out of sync with the network
+-- - If the wallet goes out of sync with the node
+monitorStatus :: Tracer IO (FollowLog tr) -> IO (Tracer IO (FollowLog tr))
+monitorStatus tr = do
```
Metrics averaged over time will be good. We can export them with ekg.
For this ticket though, we just want to log interesting events and state changes.

It _may_ work better if log interpretation and transformation is done at the app top-level, all together, rather than inside each component. I'm not sure yet though.

Anviking replied:
> It may work better if log interpretation and transformation is done at the app top-level, all together, rather than inside each component. I'm not sure yet though.

Yeah, I was wondering about that too. Partly why I wondered if iohk-monitoring was the "right way" (despite its weirdness), as I think their backends and configuration setup served exactly as that (top-level, all together transformation).

I suspect we can get by for now doing the transformation here.

Anviking replied:
> For this ticket though, we just want to log interesting events and state changes.

IMO, being able to tell 
1. If the wallet is restoring slowly
2. If the wallet is rolling back in a loop (I _think_ the stats would be able to tell this)
is an important thing we lose when we remove the "on-each-batch-logs"

rvl reviewed (commented):
Good, on the right track. I also have some related logging fixes in #2570, so I'll pull them out into a separate PR.

Anviking committed:
Move logic into follow from call-sites

Goal: have the follow function itself be responsible to handle rollbacks
and exceptions. If follow is only called /once/ designing good logs
should be easier. The follow function can then provide more information
in logs, as it knows what it will do on e.g. exceptions.

Move Tip and SyncProgress tracing to follow

We will re-add the syncProgress tracing shortly a as part of the
log aggregation

rvl commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -750,73 +752,60 @@ listUtxoStatistics ctx wid = do
         , IsOurs s RewardAccount
         )
     => ctx
+    -> Tracer IO WalletFollowLog
```
There's a tracer in `ctx` already, so no param needed.

Anviking replied:
The idea was to have `follow` provide this tracer through the callback, such that it could intercept and aggregate on this "inner" trace too.

I didn't end up doing that, but I think the concept is sound and something that could be done in the future.

rvl commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -875,23 +864,21 @@ restoreBlocks ctx wid blocks nodeTip = db & \DBLayer{..} -> mapExceptT atomicall
     logDelegation :: (SlotNo, DelegationCertificate) -> IO ()
-    logDelegation (slotNo, cert) = traceWith tr $ MsgDelegation slotNo cert
+    logDelegation (slotNo, cert) = traceWith tr
+        $ MsgDiscoveredDelegationCert slotNo cert
```
```suggestion
    logDelegation = traceWith tr . uncurry MsgDiscoveredDelegationCert
```

rvl replied:
Delegations could be logged all together too.

rvl commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -626,7 +626,8 @@ migrateManually tr proxy defaultFieldValues =
                 value <- either (fail . show) (\x -> pure $ "\"" <> x <> "\"") $
                     fromPersistValueText (toPersistValue W.EncryptWithPBKDF2)
-                traceWith tr $ MsgManualMigrationNeeded passphraseScheme value
+                traceWith tr . MsgExpectedMigration
```
Slightly better now

Anviking replied:
Yeah, not sure what the proper way would be 🤷‍♂️

rvl reviewed (commented):
Nice, I'll review more later.

Anviking commented on lib/core/src/Cardano/Wallet/Network.hs:
```diff
@@ -381,19 +490,182 @@ instance ToText FollowLog where
+    let sp = NotResponding -- Hijacked as a default value. Should be fine.
+                           -- It appears to be otherwise unused in the codebase.
+    mvar <- newMVar $ emptyStats t0' sp
+    _ <- forkIO $ loop mvar (5 * second)
```
`forkIO` — should presumable use something like `race` to get the monitoring thread to stop if `follow` stops.

rvl commented on lib/core/src/Cardano/Wallet/Network.hs:
```diff
@@ -381,19 +490,190 @@ instance ToText FollowLog where
+    _ -> s
+
+instance ToText (FollowStats LogState) where
+    toText (FollowStats b r t sp) = syncStatus <> " " <> stats
```
This time-based summary is good for the stake pools thread, but for wallets I think it should be logging when there are incoming transactions, or when restoration is in progress, and complete silence otherwise.

I'll move this statistics code into the Cardano.Wallet.Logging module too.

Anviking replied:
> but for wallets I think it should be logging when there are incoming transactions

Just to be clear: the logs collected for aggregation are still shown or not depending on severity. It is still logging discoveries.

> or when restoration is in progress

Yes, ideally the `SyncProgress` `LogState` transitions should probably be logged immediately when they occur, not only from a periodic monitor-check. My hunch is still that regular status updates wallet workers are worth it too, though.  A summary every 5 minutes should be rare compared to other Info-logs (like request logs from polling), and alert us if we see slow or no restoration progress.

Anviking replied:
Or I guess we could  _monitor_ / poll the progress regularly without _logging_ it (+ stats), if we are still in-sync.

If the chain-following were to halt, we'd then notice, without being _required_ to log every-five-minutes status updates.

rvl reviewed (commented):
Looks good. I'll apply a couple clean ups then merge.

Anviking commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -3019,21 +3018,21 @@ instance IsServerError (Request, ServerError) where
 -- may not be associated with a particular worker thread.
 data WalletEngineLog
-    = MsgWallet (WorkerLog WalletId WalletLog)
+    = MsgWalletWorker (WorkerLog WalletId W.WalletWorkerLog)
```
Not sure if this diff is from a rebase or not 🤔

Anviking replied:
No, I think it was intentional.

Anviking commented on lib/core/src/Cardano/Wallet/Network.hs:
```diff
@@ -377,21 +490,216 @@ instance ToText tr => ToText (FollowLog tr) where
+  where
+    progressMovedBackwards = current (prog s) < prev (prog s)
+    notRestored = current (prog s) /= Ready
+    noBlocks = (current (blocksApplied s) - prev (blocksApplied s)) <= 0
```
I did this as a workaround for the problem described in the `prog :: f SyncProgress` doc  above.

But having the sync progress reported wrong (not re-calculated as the current time changes), could probably be really confusing. I should have a look at this again.

Anviking replied:
This is now fixed by making it easy to calculate syncProgress without node-connection. Now it returns `NotResponding` rather than blocking.

Anviking committed:
Aggregate FollowLog and log regular summaries

For reduced noise.

Anviking committed:
Add syncProgress function to NetworkLayer

By implementing it in the NetworkLayer we are able to have it return
NotResponding when we initially don't have a TimeInterpreter on startup.

This is useful to be able to log syncProgress even if the node is not
responding.

NOTE: For simplicity, this commit removes the distinction between the
wallet and node sync tolerances in the restore bench.

The tolerance used to be 300s for waitForNodeProgress, but now we use
the one provided by the NetworkLayer, which we set to 3600s.

Anviking committed:
Remove unused trace MsgSynced

Anviking committed:
Lower sev of Msg{Did,Will}Rollback, MsgApplyBlocks

Anviking committed:
Lower MsgStakePoolRegistration sev

Anviking committed:
Rename MsgDelegation to MsgDiscoveredDelegationCert

Anviking committed:
Lower MsgCheckpoint, MsgWalletMetadata sev

Anviking committed:
Lower sev for MsgDiscoveredTxs [] case

If empty -> Debug
otherwise -> Info

Anviking committed:
Add note for MsgBlocks

rvl commented on lib/core/src/Cardano/Wallet/Network.hs:
```diff
@@ -377,21 +480,223 @@ instance ToText tr => ToText (FollowLog tr) where
+    let tr' = flip contramapM tr $ \msg -> do
+            atomically $ do
+                s <- takeTMVar var
+                putTMVar var (updateStats msg s)
```
This is not going to leave unevaluated thunks, is it? I suppose all the results are consumed by the other thread.

Anviking replied:
Good that you brought it up again. I did add a `force` in `updateStats` long time ago, but:
1. I'm not sure if that was effective or not because the outer lazy `putTMVar`
2. After asking in #haskell it seems it seems `force` is not the right way (but ok, also sounds familiar)

Anviking replied:
I replaced `TMVar` with `StrictTMVar` for now (could have used `$!` on this line also I think).

Will have a look at how to replace the `force` with something better in a follow-up PR.

rvl reviewed (approved):
Good improvement :+1:

Anviking committed:
Rename (FollowLog tr) to (FollowLog msg)

Anviking committed:
Fixup: use StrictTMVar instead of TMVar

To be even stricter without using $!.
